quickwit/quickwit-ingest/src/ingest_v2/gc.rs

```diff
@@ -0,0 +1,225 @@
+// Copyright (C) 2023 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::Duration;
+
+use mrecordlog::error::DeleteQueueError;
+use mrecordlog::MultiRecordLog;
+use once_cell::sync::Lazy;
+use quickwit_proto::ingest::IngestV2Result;
+use quickwit_proto::metastore::{DeleteShardsRequest, DeleteShardsSubrequest};
+use quickwit_proto::{split_queue_id, IndexUid, QueueId, ShardId, SourceId};
+use tokio::sync::{RwLock, Semaphore};
+use tracing::{error, info};
+
+use super::ingest_metastore::IngestMetastore;
+use super::ingester::IngesterState;
+
+/// Period of time after which shards candidate for deletion are actually deleted.
+///
+/// A shard is candidate for deletion when it is closed, its publish position greater or equal than
+/// its replication position, and the grace period has elapsed.
+const GRACE_PERIOD: Duration = if cfg!(any(test, feature = "testsuite")) {
+    Duration::from_millis(10)
+} else {
+    Duration::from_secs(60 * 10)
+};
+
+/// Maximum number of concurrent GC tasks.
+const MAX_CONCURRENCY: usize = 3;
+
+/// Limits the number of concurrent GC tasks.
+static MAX_CONCURRENCY_SEMAPHORE: Lazy<Arc<Semaphore>> =
+    Lazy::new(|| Arc::new(Semaphore::new(MAX_CONCURRENCY)));
+
+/// Deletes shards asynchronously after the grace period has elapsed.
+///
+/// 1. deletes the shards from the metastore
+/// 2. removes the shards from in-memory data structures
+/// 3. deletes the associated mrecordlog queues
```

Comment:
I am not sure exactly how to fix this, but the structure of the code and the naming made me thing that this task was a loop.

I think I would have been less confused with a simple function that does not include the sleep in this file, and 
a call that includes the grace period sleep on the call site.

Reply:
...because it was a loop at some point :D I will package everything in a function.